### PR TITLE
service: differentiate UDP and TCP protocols

### LINKED ISSUE #9207: Differentiate between UDP and TCP services
Currently, we do not differentiate between UDP and TCP services both in the BPF datapath and cilium-agent which means that services of different L4 protocols with the same port number cannot co-exist.

Solving the issue should be thoroughly thought, as we don't want to break the ongoing connections.

### PR DESCRIPTION
**Pull request by jibi**
This PR introduces support for protocol differentiation (UDP, TCP) for services.

The high level idea (first commit) is to take into account the L4 protocol specified in a k8s service when populating the bpf lb maps, and then use the packet's protocol to lookup services in those maps.

Next, a new `--bpf-lb-proto-diff` flag (enabled by default) is introduced. When disabled, the agent strips the protocol from the service object, so even though all the protocol differentiation logic is in place, the control plane sees only services with `NONE` protocol, effectively falling back to the old behavior. In addition to that also the datapath stops taking into account the protocol when looking up the lb maps.

Next we tackle the reason why the previous attempts didn't make it into the main branch: connection disruptions during upgrades and downgrades. The proposed solution is to keep, during upgrades, existing old style/`NONE` services as such instead of deleting them and creating new protocol-aware ones. In the control plane before creating a new service, even if protocol differentiation is enabled, we check if the same service with `NONE` protocol already exists. If that's the case, we don't create a new one. In the datapath we add an extra lookup to the lb maps: if the protocol-aware lookup fails, we set the protocol to 0 (`NONE`) and try again. This allows to go through upgrades and downgrades without disrupting connections for _existing_ `NONE` services. This doesn't solve all the cases: if a service has its protocol set and Cilium is downgraded to a version that doesn't support protocol differentiation, such connection will break as the service will be recreated.

Forth commit is just about changing `NONE` to `ANY` when dealing with no-proto services.

Finally we update all the unit and ginkgo tests to make them aware of the protocol differentiation logic, as well as the CLI to make a test aware of the protocol differentiation.

I'm currently looking into additional unit tests for controlplane and datapath, as well as additional documentation, but given that the functionality is in this PR should be already good for a first round of reviews.

### CODE REVIEW

File: pkg/loadbalancer/loadbalancer.go
Comment (aojea) [inline pkg/loadbalancer/loadbalancer.go:591]: self note, it is a bit late but this will be better called "any"

File: pkg/service/service.go
Comment (aojea) [inline pkg/service/service.go:1423]: this deserves a comment explaining the logic here and is not clear to me without having to understand the whole code base:
1. we store the original proto
2. we mutate everything to NONE
3. if does not exist with NONE with try with the original proto

did I get it right?

File: Documentation/cmdref/cilium-agent.md
Comment (aojea) [inline Documentation/cmdref/cilium-agent.md]: nit, naming bikeshedding 

`--enable-service-protocol-awareness   Enable support for service protocol awareness (TCP, UDP, SCTP) ` ??

File: daemon/cmd/daemon_main.go
Comment (aojea) [inline daemon/cmd/daemon_main.go]: I'm not familiar with cilium config options so take this comment with that in mind.
I wonder if you want to make it the negative `--disable-service-protocol-awareness` with default `false`, users will automatically pick this value without the need of setting the flag

File: pkg/k8s/watchers/service.go
Comment (aojea) [inline pkg/k8s/watchers/service.go:505]: you have `oldSvc` and `svc`, wonder if you need to modify the `oldSvc` too

File: pkg/k8s/watchers/service.go
Comment (aojea) [inline pkg/k8s/watchers/service.go:223]: at one point you may not need to do this as all services will be migrated so you may want to use the feature flag?

File: pkg/loadbalancer/loadbalancer.go
Comment (aojea) [reply pkg/loadbalancer/loadbalancer.go:591]: lol, I commented before getting to this commit :)

File: daemon/cmd/daemon_main.go
Comment (jibi) [reply daemon/cmd/daemon_main.go]: > users will automatically pick this value without the need of setting the flag

that's already the case thanks to that second parameter set to `true`, which is the default value for the flag :+1:

File: pkg/k8s/watchers/service.go
Comment (jibi) [reply pkg/k8s/watchers/service.go:223]: I think putting this behind a feature flag would prevent deleting "old" (ANY) services after an upgrade.

Here's an example:
* user is running v1.15, so no support for service proto differentiation, and has a service deployed
* user upgrades to a Cilium version that supports service proto differentiation, and enables it (or better, doesn't explicitly disable it) 
* after the upgrade, the existing service is maintained as ANY
* at some point the user deletes the service

if this logic was put behind the feature flag, we would never be able to delete that old ANY service from Cilium state/maps, as we would only try to delete the service with the protocol set

File: pkg/k8s/watchers/service.go
Comment (jibi) [reply pkg/k8s/watchers/service.go:505]: yep, need to look better into this :+1:

File: pkg/service/service.go
Comment (jibi) [reply pkg/service/service.go:1423]: that's correct :+1: 

> this deserves a comment explaining the logic here

yes, I'm planning to sprinkle some extra comments once I start cleaning up all this WIP commits

File: daemon/cmd/daemon_main.go
Comment (aojea) [reply daemon/cmd/daemon_main.go]: I see, it always pick the default you set from the flag

File: pkg/k8s/watchers/service.go
Comment (aojea) [reply pkg/k8s/watchers/service.go:223]: I worded it wrong, my point is that you only run this if `enable-service-protocol-differentiation` is set , that is the feature flag ... if in a long future there is no need for setting `enable-service-protocol-differentiation` because all cilium clusters are known to have been migrated, you no longer need to do this ... is just a comment

Comment (jibi) [reply pkg/k8s/watchers/service.go:223]: ah sorry, I see what you mean now, yes, I need to add also some comments to guide us through all the code paths that need be removed once/if `enable-service-protocol-differentiation` will be deprecated :+1:

File: Documentation/cmdref/cilium-agent.md
Comment (jibi) [reply Documentation/cmdref/cilium-agent.md]: I think I'll keep the current one as it's also in line with the original gh issue naming

File: pkg/k8s/watchers/service.go
Comment (jibi) [reply pkg/k8s/watchers/service.go:505]: > wonder if you need to modify the oldSvc too

yes, if we want to compare apples to apples `oldSvc` needs to go through `stripServiceProtocol` in tandem with `svc` :+1:

**Review by lmb — APPROVED** [review]
Loader changes LGTM.

**Comment by borkmann** [comment]
[...]
> Next, a new `--enable-service-protocol-differentiation` flag (enabled by default) is introduced. When disabled, the agent strips the protocol from the service object, so even though all the protocol differentiation logic is in place, the control plane sees only services with `NONE` protocol, effectively falling back to the old behavior. In addition to that also the datapath stops taking into account the protocol when looking up the lb maps.

nit: Could we change the flag name into `--bpf-lb-proto-diff` or such? At least we have most LB-related flags prefixed under `bpf-lb-...`, would be good to keep this consistent.

> Next we tackle the reason why the previous attempts didn't make it into the main branch: connection disruptions during upgrades and downgrades. The proposed solution is to keep existing old style/`NONE` services as such instead of deleting them and creating new protocol-aware ones. In the control plane before creating a new service, even if protocol differentiation is enabled, we check if the same service with `NONE` protocol already exists. If that's the case, we don't create a new one. In the datapath we add an extra lookup to the lb maps: if the protocol-aware lookup fails, we set the protocol to 0 (`NONE`) and try again. This allows to go through upgrades and downgrades without disruption connections for _existing_ `NONE` services. This doesn't solve all the cases: if a service has its protocol set and Cilium is downgraded to a version that doesn't support protocol differentiation, such connection will break as the service will be recreated.
> 
> Forth commit is just about changing `NONE` to `ANY` when dealing with no-proto services.

Quick q: I was wondering if there is a difference between `NONE` vs `ANY` in your description, so once this lands `ANY` will be used as opposed to `NONE`, right? In other words, there won't be a `NONE` visible to the user? (If not, could we make this consistent to just `ANY`?)

**Comment by jibi** [comment]
> [...]
> 
> > Next, a new `--enable-service-protocol-differentiation` flag (enabled by default) is introduced. When disabled, the agent strips the protocol from the service object, so even though all the protocol differentiation logic is in place, the control plane sees only services with `NONE` protocol, effectively falling back to the old behavior. In addition to that also the datapath stops taking into account the protocol when looking up the lb maps.
> 
> nit: Could we change the flag name into `--bpf-lb-proto-diff` or such? At least we have most LB-related flags prefixed under `bpf-lb-...`, would be good to keep this consistent.

sure, I'll rename that :+1: 

> > Next we tackle the reason why the previous attempts didn't make it into the main branch: connection disruptions during upgrades and downgrades. The proposed solution is to keep existing old style/`NONE` services as such instead of deleting them and creating new protocol-aware ones. In the control plane before creating a new service, even if protocol differentiation is enabled, we check if the same service with `NONE` protocol already exists. If that's the case, we don't create a new one. In the datapath we add an extra lookup to the lb maps: if the protocol-aware lookup fails, we set the protocol to 0 (`NONE`) and try again. This allows to go through upgrades and downgrades without disruption connections for _existing_ `NONE` services. This doesn't solve all the cases: if a service has its protocol set and Cilium is downgraded to a version that doesn't support protocol differentiation, such connection will break as the service will be recreated.
> > Forth commit is just about changing `NONE` to `ANY` when dealing with no-proto services.
> 
> Quick q: I was wondering if there is a difference between `NONE` vs `ANY` in your description, so once this lands `ANY` will be used as opposed to `NONE`, right? In other words, there won't be a `NONE` visible to the user? (If not, could we make this consistent to just `ANY`?)

both should be just a string representation for protocol `0`, so the user should not see any `NONE` protocol, but I'll double check that's the case in all the different scenarios

**Review by christarazi — APPROVED** [review]
LGTM for ci-structure

**Review by nebril — APPROVED** [review]
helm OK

**Comment by aojea** [comment]
❤️

**Review by nathanjsweet — APPROVED** [review]
Very clever.

File: cilium-dbg/cmd/service_update.go
Comment (derailed) [inline cilium-dbg/cmd/service_update.go:71]: Could be out of context here but do we need to handle "sctp"?

**Review by derailed — APPROVED** [review]
@jibi Nice work!

File: .github/workflows/conformance-e2e.yaml
Comment (viktor-kurchenko) [inline .github/workflows/conformance-e2e.yaml]: Is this temporary change?

File: cilium-dbg/cmd/service_update.go
Comment (jibi) [reply cilium-dbg/cmd/service_update.go:71]: because that's part of the k8s supported protocols: https://kubernetes.io/docs/reference/networking/service-protocols/

File: .github/workflows/conformance-e2e.yaml
Comment (jibi) [reply .github/workflows/conformance-e2e.yaml]: yes, see the commit (https://github.com/cilium/cilium/pull/33434/commits/0bef665c968bec7b5324fc27c78cb3065b8eada1) title for that change

**Review by viktor-kurchenko — APPROVED** [review]

**Review by ghost — APPROVED** [review]
docs good!

**Review by cagataygurturk — APPROVED** [review]

**Review by borkmann — APPROVED** [review]

**Comment by ldelossa** [comment]
@jibi this PR is marked ready to merge, but GH is complaining it still needs a review from sig/loader. Can you address?

**Comment by jibi** [comment]
> @jibi this PR is marked ready to merge, but GH is complaining it still needs a review from sig/loader. Can you address?

I think it's because Lorenz reviewed it but he is not in sig/loader anymore. But anyway, I need to fix one more CI failure, so it's definitely not ready to be merged

**Review by dylandreimerink — APPROVED** [review]
Loader changes LGTM

**Comment by ldelossa** [comment]
@jibi looks like Dylan's review pushed the ready-to-merge label back on. I don't see a force push or anything since your last comment, so removing the label.

**Review by sayboras — APPROVED** [review]
Thanks a lot for your help tracking down the upgrade failure in Ingress.

**Comment by julianwiedmann** [comment]
> Next, a new `--bpf-lb-proto-diff` flag (enabled by default) is introduced. When disabled, the agent strips the protocol from the service object, so even though all the protocol differentiation logic is in place, the control plane sees only services with `NONE` protocol, effectively falling back to the old behavior. In addition to that also the datapath stops taking into account the protocol when looking up the lb maps.

Can we already discuss what a deprecation schedule for this flag would look like? :)

**Comment by jibi** [comment]
> Can we already discuss what a deprecation schedule for this flag would look like? :)

we could deprecate it in v1.18 and remove it completely in 1.19 :thinking: 
different story for the backward compatibility logic for services with protocol=ANY, that could potentially stay there for a long time (as we have no way to control whether users have recreated their services/there are no more protocol=ANY services around)

**Comment by julianwiedmann** [comment]
> > Can we already discuss what a deprecation schedule for this flag would look like? :)
> 
> we could deprecate it in v1.18 and remove it completely in 1.19 🤔 different story for the backward compatibility logic for services with protocol=ANY, that could potentially stay there for a long time (as we have no way to control whether users have recreated their services/there are no more protocol=ANY services around)

Tracking this deprecation here: https://github.com/cilium/cilium/issues/39249

**Pull request merged**
